Refactor: RandomActivity to use ViewModel

Lampotrias opened the pull request:
Hi all!

It's my version refactoring RandomActivity with ViewModel.
https://phabricator.wikimedia.org/T303942

Thanks in advance for the code review!

dbrant commented on app/src/main/java/org/wikipedia/random/RandomViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+		L.w(throwable)
+	}
+
+	private val _savaShareState = MutableStateFlow(false)
```
misspelling

dbrant commented on app/src/main/java/org/wikipedia/random/RandomViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+	}
+
+	private val _savaShareState = MutableStateFlow(false)
+	val savaShareState: StateFlow<Boolean> = _savaShareState
```
misspelling

dbrant commented on app/src/main/java/org/wikipedia/random/RandomViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+	private val _savaShareState = MutableStateFlow(false)
+	val savaShareState: StateFlow<Boolean> = _savaShareState
+
+	private val _saveToDefaultList = MutableStateFlow<SaveSharedState<PageTitle>>(Initial())
```
What's the reason for having underscore `_` and non-underscore versions of all these fields?

Lampotrias replied:
The non-underscored fields used only observe values in other entities, such fragments or activities. The underscored field guarantees emit values only from view model. Unfortunately now not other way for it.

dbrant commented on app/src/main/java/org/wikipedia/random/RandomViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+		funnel = RandomizerFunnel(WikipediaApp.getInstance(), wikiSite, invokeSource)
+	}
+
+	fun saveToDefaultList(title: PageTitle) {
+		_saveToDefaultList.value = Result(title)
+	}
+
+	fun saveToCustomList(title: PageTitle) {
+		_saveToCustomList.value = Result(title)
+	}
+
+	fun movePageToList(sourceReadingListId: Long, title: PageTitle) {
+		_movePageToList.value = Result(MoveDataSource(sourceReadingListId, title))
+	}
```
I'm not sure I see the benefit of the above three functions. You're passing in a PageTitle, which gets passed to a Flow, which gets passed right back to the Fragment. Why not keep these interactions within the Fragment as before?

Lampotrias replied:
I'll fix it. I created these methods with the expectation that the logic could be complicated in the future

Lampotrias commented:
Thank you for your feedback.

I fixed all defects in pull request. In RandomItemFragment was added viewmodel and removed `retainInstance = true`. I tested case with rotation of screen and saw no problem with it.

Lampotrias commented on app/src/main/java/org/wikipedia/random/RandomItemViewModel.kt:
```diff
@@ -0,0 +1,58 @@
+	private val wikiSite: WikiSite = bundle.getParcelable(RandomActivity.INTENT_EXTRA_WIKISITE)!!
+
+	private val _saveShareState = MutableStateFlow<SaveSharedState>(Initial)
+	val saveShareState = _saveShareState.asStateFlow()
```
Yes, you are right.

Lampotrias commented on app/src/main/java/org/wikipedia/random/RandomItemFragment.kt:
```diff
@@ -33,7 +34,7 @@ class RandomItemFragment : Fragment() {
-    private val disposables = CompositeDisposable()
+    private val viewModel: RandomItemViewModel by viewModels { RandomItemViewModel.Factory(requireArguments()) }
 
     private lateinit var wikiSite: WikiSite
```
Now, I don't need `wikiSite`  in the ViewModel, I planning use `WikipediaApp.getInstance().wikiSite`. Currently, `wikiSite` use in `RandomItemFragment` for call:
```L10nUtil.setConditionalLayoutDirection(view, wikiSite.languageCode)```, as it was before the refactoring

Lampotrias replied:
Ok, now `wikiSite` is given from `viewModel`.

Lampotrias commented on app/src/main/java/org/wikipedia/random/RandomItemViewModel.kt:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	fun getRandomPage() {
+		val d = ServiceFactory.getRest(wikiSite).randomSummary
```
I have question with select MediaWiki Action API. Do I need to make a request like `https://en.wikipedia.org/w/api.php?action=query&format=json&list=random&formatversion=2` and then get an instance of PageSummary

Lampotrias commented:
Thank you for your feedback.

I removed all code, that was created for the future, such Factory.

And sorry about the copy-paste code. This is my first PR, I'm a little worried.
